### DISCUSSION TITLE
Exception during Context constructor

### DISCUSSION THREAD

**Discussion by nathangiuliani** [cake-build/cake#2937] [Q&A] [answered]
Hi again,

While working on moving some build processes to Frosting, I've found that exceptions thrown during the construction of the Context class aren't handled well - the only output when executing outside of an IDE is `Error: Exception has been thrown by the target of an invocation.`. This can be easily reproduced by modifying the `Context.cs` file on a new frosting project deployed from the template:

```
using Cake.Core;
using Cake.Frosting;

public class Context : FrostingContext
{
    public Context(ICakeContext context) 
        : base(context)
    {
        throw new System.Exception();
    }
}
```

At this point my workaround is to handle the exception and some useful output within the constructor, however I think that perhaps the issue here might be that not using this as intended...

At this point our context class - which is the context class from the template - has additional properties for all of our build variables, mostly set from command line arguments, but some things are calculated or set from other tools. In this case, the issue was that GitVersion could not be found, because the tools installation process doesn't run until after the context constructor is called. 

So, what's the right way to go about this?

Thanks

**Comment by nils-a** [comment, marked as answer]
Generally setup should be done in a `IFrostingLifetime.Setup`.

Somthing like
```cake
public sealed class Lifetime : FrostingLifetime<Context>
{
    public override void Setup(Context context)
    {
        context.Information("Setting things up...");
        // here goes the setup..
    }

    public override void Teardown(Context context, ITeardownContext info)
    {
        context.Information("Tearing things down...");
    }
}
```
